main.go

```diff
@@ -70,63 +58,169 @@ func (m model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
 	switch msg := msg.(type) {
 	case tea.KeyMsg:
 		switch msg.Type {
-		case tea.KeyCtrlC, tea.KeyEsc: // leave typing mode or quit if out
-			if m.typing {
-				m.Stop()
-			} else {
+		case tea.KeyCtrlC: // quit
+			return m, tea.Quit
+		case tea.KeyEsc:
+			switch m.state {
+			case ViewTitle:
 				return m, tea.Quit
+			case ViewTest:
+				m.CalculateStats()
+				m.state = ViewResults
 			}
 		case tea.KeyBackspace:
-			if m.typing && len(m.typed) > 0 {
+			if m.state == ViewTest && len(m.typed) > 0 {
 				m.typed = m.typed[:len(m.typed)-1]
 			}
 		case tea.KeyEnter:
-			m = initialModel()
+			switch m.state {
+			case ViewTitle:
+				// begin new test
+				m.state = ViewTest
+				m.start = time.Now()
+				m.typed = ""
+			case ViewTest:
+				// restart test
+				m.typed = ""
+				m.start = time.Now()
+			case ViewResults:
+				// begin a new test
+				m.state = ViewTest
+				m.start = time.Now()
+				m.typed = ""
+			}
 		default:
-			m.Start()
-			m.typed += string(msg.Runes)
-
-			if m.typed == m.text {
-				m.Finish()
+			if m.state == ViewTest {
+				m.typed += string(msg.Runes)
+				if m.ShouldEndTest() {
+					m.CalculateStats()
+					m.state = ViewResults
+				}
 			}
 		}
+	case tea.WindowSizeMsg:
+		m.width = msg.Width
+		m.height = msg.Height
 	}
 
 	return m, nil
 }
 
-func colorText(text, color string) string {
-	return fmt.Sprintf("%s%s%s", color, text, colorReset)
+func (m *model) CalculateStats() {
+	m.lastDuration = time.Since(m.start)
+	durationInMinutes := m.lastDuration.Minutes()
+	m.statistics.cpm = int(float64(len(m.typed)) / durationInMinutes)
+	m.statistics.wpm = int(float64(len(m.typed)) / 5 / durationInMinutes)
+}
+
+func (m model) ShouldEndTest() bool {
+	textWords := strings.Fields(m.text)
+	typedWords := strings.Fields(m.typed)
+
+	// if typed more words end test
+	if len(typedWords) > len(textWords) {
+		return true
+	}
+
+	// all words typed
+	if len(typedWords) == len(textWords) {
+
```

Comment:
We should create some variables 
```go
lastTypedWord := typedWords[len(typedWords)-1]
lastTextWord := textWords[len(textWords)-1]
```

This will help with the readability of the later if conditions